### ISSUE TITLE
upnpc -U: updating pinholes depends on CheckPinholeWorking being available

### ISSUE THREAD

**Issue by stefantalpalaru** [miniupnp/miniupnp#658]
I have a Fritzbox 7590 that does not provide a "CheckPinholeWorking" action. Tested using [async-upnp-client](https://pypi.org/project/async-upnp-client/):

```text
$ upnp-client --pprint call-action http://192.168.1.1:49000/igd2desc.xml WANIPv6Firewall1/CheckPinholeWorking UniqueID=4
[...]
async_upnp_client.exceptions.UpnpActionResponseError: Error during async_call(), status: 500, upnp error: 403 (Not available Action)
```

`upnpc` gives me the same 403 error code, when trying to update a pinhole, because it insists on checking that it's working first:

```text
$ upnpc -U 4 3600
upnpc : miniupnpc library test client, version 2.2.4.
 (c) 2005-2022 Thomas Bernard.
Go to http://miniupnp.free.fr/ or https://miniupnp.tuxfamily.org/
for more information.
List of UPNP devices found on the network :
 desc: http://192.168.1.1:49000/igddesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://192.168.1.1:49000/igdupnp/control/WANIPConn1
Local LAN ip address : 192.168.1.8
CheckPinholeWorking: Pinhole ID = 4 / IsWorking = No
CheckPinholeWorking() failed with code 403 (UnknownError)
```

So for Fritzbox, the 403 error code means that the action is not available, but I'm not sure it means the same thing for all modem brands. All 4xx codes are errors, of course, but [this Microsoft document](https://learn.microsoft.com/en-us/windows/win32/upnp/device-error-codes) lists 403 as "UPNP\_E\_OUT\_OF\_SYNC".

Back to `upnpc`.  When asked to update a pinhole's lease time with `-U`, it will only do it after running `PNP_CheckPinholeWorking()` which must be either successful or return the error code 709 (meaning "No traffic corresponding to this pinhole has been received by the gateway." according to the [official docs](file:///mnt/media_no_backup/downloads/UPnP-gw-WANIPv6FirewallControl-v1-Service-1.pdf)): https://github.com/miniupnp/miniupnp/blob/f91a32ff02769701baa1f46c5751c0eb1b8f2f8d/miniupnpc/src/upnpc.c#L468-L478

But the same official docs specify that "CheckPinholeWorking" is an optional action. Manufacturers don't have to implement it, to be compatible with the standard.

So, while we could add error code 403 as an exception (along the existing 709), the better fix is to completely remove the "working" check before trying to update the pinhole.

**Comment by miniupnp** [comment]
I think there is and issue with the Fritzbox 7590 implementation, the UPnP error 403 is deprecated since [UPnP Device Architecture v1.0](http://www.upnp.org/specs/arch/UPnP-arch-DeviceArchitecture-v1.0-20080424.pdf) (see page 53).

If it does not implements an optional action, it should return error 602 : 
602 / Optional Action Not Implemented / The requested action is optional and is not implemented by the device.

**Comment by stefantalpalaru** [comment]
> I think there is and issue with the Fritzbox 7590 implementation

Probably, but we deal with actual implementations, not ideal ones. I still think the update should be tried unconditionally.

**Comment by miniupnp** [comment]
should be fixed with 51fa714bbbf476c4d5de8ad0e32c9e344e9074a5 and miniupnpc version 2.2.5
